atlas_splitter/barrel_splitter/somatosensory_barrels.py

```diff
@@ -0,0 +1,294 @@
+"""
+Algorithm introducing volumes of 33 barrel columns coming from segmentation of the
+average brain atlas images from Allen institute. Barrel columns are introduced as children
+to somatosensory area of barrel cortex and then subdivided into layers. Each barrel is
+subdivided into layers. To account for this being used after `layer-splitter` layer 2/3 is
+split into 2 and 3.
+
+* Introduction of barrels to the hierarchy.json
+* Introduction of the annotated volumes to the annotations.nrrd
+
+The code relays on pd.DataFrame containing the annotated positions of each barrel voxel
+in x,y,z coordinates.
+"""
+import copy
+import logging
+from collections import defaultdict
+from typing import Any, Dict, Iterator, List
+
+import numpy as np
+import pandas as pd
+from voxcell import RegionMap, VoxelData
+
+from atlas_splitter.utils import _assert_is_leaf_node, create_id_generator, get_isocortex_hierarchy
+
+L = logging.getLogger(__name__)
+HierarchyDict = Dict[str, Any]
+
+
+def layer_ids(
+    id_generator: Iterator[int], names: List[str], layers: List[str]
+) -> Dict[str, Dict[str, int]]:
+    """Create a dictionary of ids for the new regions with layer subregions.
+
+    Args:
+        id_generator: Iterator that generates new ids.
+        names: A list of names of the new regions.
+        layers: A list of names of the new layer regions.
+
+    Returns:
+        A dictionary that maps each region to a dictionary of its layer subregions and their ids.
+    """
+    new_ids: Dict[str, Dict[str, int]] = defaultdict(dict)
+
+    for name in names:
+        new_ids[name] = {}
+        new_ids[name][name] = next(id_generator)
+
+        for layer_name in layers:
+            new_ids[name][layer_name] = next(id_generator)
+
+    return new_ids
+
+
+def get_hierarchy_by_acronym(hierarchy: HierarchyDict, acronym: str):
+    """Find and return child with a matching acronym from the next level
+    of the hierarchy.
+
+    Args:
+        hierarchy (HierarchyDict): brain regions hierarchy dict
+        acronym (str): name of the region
+
+    Returns:
+        HierarchyDict: HierarchyDict of the matching child
+    """
+    for index, child in enumerate(hierarchy["children"]):
+        if child["acronym"] == acronym:
+            return hierarchy["children"][index]
+
+    return None
+
+
+def positions_to_mask(positions: np.ndarray, annotation: VoxelData) -> np.ndarray:
+    """Change x, y, z position coordinates into binary mask in 3D boolean array.
+
+    Args:
+        positions: A numpy array of shape (n, 3) representing the x, y, z positions.
+        annotation: A VoxelData object representing the orientation field of the atlas.
+
+    Returns:
+        A numpy array of shape annotation.shape with True values at the indices
+        corresponding to the input positions.
+    """
+    mask = np.zeros(annotation.shape, dtype=bool)
+    indices = annotation.positions_to_indices(positions)
+    for indx in indices:
```

Comment:
```suggestion
    for indx in indices:
```

this for loop can be replaced w/ `mask[tuple(indices.T)] = True` for a large speedup